packages/react-reconciler/src/SchedulingProfiling.js

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {Lane, Lanes} from './ReactFiberLane';
+import type {Fiber} from './ReactInternalTypes';
+
+import {enableSchedulingProfiling} from 'shared/ReactFeatureFlags';
+import {getStackByFiberInDevAndProd} from './ReactFiberComponentStack';
+
+/**
+ * If performance exists and supports the subset of the User Timing API that we
+ * require.
+ */
+const supportsUserTiming =
+  typeof performance !== 'undefined' && typeof performance.mark === 'function';
+
+function formatLanes(laneOrLanes: Lane | Lanes): string {
+  return ((laneOrLanes: any): number).toString();
+}
+
+export function markCommitStarted(lanes: Lanes): void {
+  if (enableSchedulingProfiling) {
+    if (supportsUserTiming) {
+      performance.mark(`--commit-start-${formatLanes(lanes)}`);
+    }
+  }
+}
+
+export function markCommitStopped(): void {
+  if (enableSchedulingProfiling) {
+    if (supportsUserTiming) {
+      performance.mark(`--commit-stop`);
+    }
+  }
+}
+
+const PossiblyWeakMap = typeof WeakMap === 'function' ? WeakMap : Map;
+// $FlowFixMe: Flow cannot handle polymorphic WeakMaps
+const wakeableIDs: WeakMap<Wakeable, number> = new PossiblyWeakMap();
+let wakeableID: number = 0;
+function getWakeableID(wakeable: Wakeable): number {
+  if (!wakeableIDs.has(wakeable)) {
+    wakeableIDs.set(wakeable, wakeableID++);
+  }
+  return ((wakeableIDs.get(wakeable): any): number);
+}
+
+export function markComponentSuspended(
+  componentName: string,
+  fiber: Fiber,
+  wakeable: Wakeable,
+): void {
+  if (enableSchedulingProfiling) {
+    if (supportsUserTiming) {
+      const id = getWakeableID(wakeable);
+      const componentStack = getStackByFiberInDevAndProd(fiber) || '';
+      // TODO (brian) Generate and store temporary ID so DevTools can match up a component stack later.
```

Comment:
I like Dan's suggestion of caching component stacks (using a weak map) for fibers we've seen.

Let's move these TODO comments above the calls to `getStackByFiberInDevAndProd` since it relates to those calls.

The idea behind the "TODO" was that- once this functionality is more tightly integrated into DevTools- try to store a (weak) map of id-to-Fiber here, and then let DevTools compute the component stack asynchronously later, using the Fiber. I'm not sure if that is feasible though, especially with future plans to replace the `alternate` model.

Reply:
Btw we have something called `fiber._debugID`. Maybe we can repurpose this?

Reply:
That sounds like it could work.

I'm not sure what the implications of the alternate -> previous change would be for this idea. It was kind of a half thought out idea anyway. Just something I wanted to put some thought into later because of the cost of getting the component stack.

Component stack calculation has gotten even more expensive recently with the "native" component stacks so... I don't know how I feel about this at the moment. It's definitely useful info to have, but maybe it's no longer worth the cost?

@sebmarkbage may have an opinion.